crates/bevy_animation/src/lib.rs

```diff
@@ -479,173 +539,99 @@ impl AnimationPlayer {
-        return None;
-    };
-    if names.get(current_entity) != Ok(root_name) {
-        return None;
-    }
-
-    for (idx, part) in parts {
-        let mut found = false;
-        let children = children.get(current_entity).ok()?;
-        if let Some(cached) = path_cache[idx] {
-            if children.contains(&cached) {
-                if let Ok(name) = names.get(cached) {
-                    if name == part {
-                        current_entity = cached;
-                        found = true;
-                    }
-                }
-            }
-        }
-        if !found {
-            for child in children.deref() {
-                if let Ok(name) = names.get(*child) {
-                    if name == part {
-                        // Found a children with the right name, continue to the next part
-                        current_entity = *child;
-                        path_cache[idx] = Some(*child);
-                        found = true;
-                        break;
-                    }
-                }
-            }
-        }
-        if !found {
-            warn!("Entity not found for path {:?} on part {:?}", path, part);
-            return None;
+/// A system that advances the time for all playing animations.
+pub fn advance_animations(
+    time: Res<Time>,
+    animation_clips: Res<Assets<AnimationClip>>,
+    mut players: Query<&mut AnimationPlayer>,
+) {
+    for mut player in players.iter_mut() {
+        let paused = player.paused;
+        if paused {
+            continue;
         }
-    }
-    Some(current_entity)
-}
 
-/// Verify that there are no ancestors of a given entity that have an [`AnimationPlayer`].
-fn verify_no_ancestor_player(
-    player_parent: Option<&Parent>,
-    parents: &Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-) -> bool {
-    let Some(mut current) = player_parent.map(Parent::get) else {
-        return true;
-    };
-    loop {
-        let Ok((has_player, parent)) = parents.get(current) else {
-            return true;
+        // Advance the main animation.
+        if let Some(animation_clip) = animation_clips.get(&player.animation.animation_clip) {
+            player
+                .animation
+                .update(time.delta_seconds(), animation_clip.duration);
         };
-        if has_player {
-            return false;
-        }
-        if let Some(parent) = parent {
-            current = parent.get();
-        } else {
-            return true;
+
+        // Advance transition animations.
+        for transition in &mut player.transitions {
+            // Decrease weight. But don't expire the transition yet if the time
+            // ran out; we still need to update the bones.
+            transition.current_weight = (transition.current_weight
+                - transition.weight_decline_per_sec * time.delta_seconds())
+            .max(0.0);
+
+            if let Some(animation_clip) = animation_clips.get(&transition.animation.animation_clip)
```

Comment:
It feels like we shouldn't need to re-query the asset each time we tick an animation. Can't we cache `animation_clip.duration` if that's the only thing we need?

Reply:
I usually prefer not to denormalize data unless there's a compelling performance reason to, because it leads to a lot of complexity. Assets are mutable in Bevy: what if you change the duration of the animation clip? Do we have a separate system that goes and updates all the animation players?

I think I'd prefer to do that as a follow-up if possible.